aptos-move/framework/aptos-framework/sources/Coin.move

```diff
@@ -0,0 +1,342 @@
+/// This module provides the foundation for typesafe Coins.
+module AptosFramework::Coin {
+    use Std::ASCII;
+    use Std::Errors;
+    use Std::Event::{Self, EventHandle};
+    use Std::Option::{Self, Option};
+    use Std::Signer;
+
+    use AptosFramework::TypeInfo;
+
+    const ECOIN_INFO_ADDRESS_MISMATCH: u64 = 0;
+    const ECOIN_INFO_ALREADY_PUBLISHED: u64 = 1;
+    const ECOIN_INFO_NOT_PUBLISHED: u64 = 2;
+    const ECOIN_STORE_ALREADY_PUBLISHED: u64 = 3;
+    const ECOIN_STORE_NOT_PUBLISHED: u64 = 4;
+    const EINSUFFICIENT_BALANCE: u64 = 5;
+    const ENO_BURN_CAPABILITY: u64 = 6;
+    const ENO_MINT_CAPABILITY: u64 = 7;
+
+    // Core data structures
+
+    // Represents a set amount of coin
+    struct Coin<phantom CoinType> has store {
+        value: u64,
+    }
+
+    // Represents ownership of coin
+    struct CoinStore<phantom CoinType> has key {
+        coin: Coin<CoinType>,
+        deposit_events: EventHandle<DepositEvent>,
+        withdraw_events: EventHandle<WithdrawEvent>,
+    }
+
+    // Represents resources associated with a coin
+    struct CoinInfo<phantom CoinType> has key {
+        name: ASCII::String,
```

Comment:
As names are not unique, pretty sure some smart contract dev would do something like this in his smart contract:

```
let symbol = Coin::name<T>();
if (symbol == b"USDT") { do something important };
```

And leave security hole for attackers.

Reply:
Hmm... this is an interesting point, but that's why they should be using the CoinType and not those fields. Name is just there for metadata just like scaling factor.

Reply:
Right, yet it could happen indeed because of misunderstanding generic concept and experience with other smart contracts language imho.

Anyway, it's good to know and maybe good topic to discuss in the future

Reply:
Absolutely, the contract isn't locked into place, we need to translate the concept into a standard and have more conversation around these pieces.

As you've probably seen we have both Coin.move and Token.move. This is exploratory work to try to get it right :)